src/cargo/util/toml/mod.rs

```diff
@@ -1006,9 +1012,14 @@ impl TomlManifest {
         let publish = match publish {
             Some(VecStringOrBool::VecString(ref vecstring)) => Some(vecstring.clone()),
             Some(VecStringOrBool::Bool(false)) => Some(vec![]),
-            None | Some(VecStringOrBool::Bool(true)) => None,
+            Some(VecStringOrBool::Bool(true)) => None,
+            None => version.is_none().then_some(vec![]),
         };
 
+        if version.is_none() && publish != Some(vec![]) {
+            bail!("`package.publish` requires `package.version` be specified");
+        }
```

Comment:
This always bugs me. Now it looks like

* `publish = true` -> `None`
* not specified -> `[]`

which is a good thing as it exactly aligns to the logic here:

https://github.com/rust-lang/cargo/blob/87f4b1ba07f698e0a30d75ea0c9fbdb1e8655079/src/cargo/core/package.rs#L170-L175

I wonder if we could have a type wrapper to make the logic more clear with some methods.

Reply:
We likely should refactor this to use a custom enum, rather than `Option`.